internal/dcache/rpc/server/handler.go

```diff
@@ -1773,18 +1854,40 @@ func (h *ChunkServiceHandler) JoinMV(ctx context.Context, req *models.JoinMVRequ
 	mvInfo := rvInfo.getMVInfo(req.MV)
 	if mvInfo != nil {
 		//
-		// TODO: Till Sourav formally implements idempotent handling of JoinMV and UpdateMV RPCs,
-		//	 we have the following to not treat "double join" as failure.
-		//	 Double join can happen when let's say we have two or more outofsync component RVs
-		//	 for an MV and fixMV() sends JoinMV request to each of the outofsync RVs. If one or
-		//	 more of these fail, the joinMV() will treat it as a failure and not update clustermap.
-		//	 Next time when fixMV() is called it'll again attempt fixing and again send JoinMV.
-		//	 Note that for proper handling we need to ensure that the reservedSpace remains
-		//	 same across both calls. Also if an RV is joined but never used later (maybe joinMV()
-		//	 picked a new RV in the next iteration), we should time out and undo the reservedSpace.
+		// JoinMV and UpdateMV need to be idempotent to not treat "double join" as failure.
+		// Double join can happen when let's say we have two or more outofsync component RVs
+		// for an MV and fixMV() sends JoinMV request to each of the outofsync RVs. If one or
+		// more of these fail, the joinMV() will treat it as a failure and not update clustermap.
+		// Next time when fixMV() is called it'll again attempt fixing and again send JoinMV.
+		// Note that for proper handling we need to ensure that the reservedSpace remains
+		// same across both calls. Also if an RV is joined but never used later (maybe joinMV()
+		// picked a new RV in the next iteration), we should time out and undo the reservedSpace.
+		// This one is a TODO.
+		//
+		errStr := fmt.Sprintf("Double join for %s/%s, prev join at: %s, by: %s",
+			req.RVName, req.MV, mvInfo.joinedWhen, mvInfo.joinedBy)
+
+		log.Warn("ChunkServiceHandler::JoinMV: %s", errStr)
+
 		//
-		log.Warn("ChunkServiceHandler::JoinMV: %s is already part of %s, ignoring", req.RVName, req.MV)
-		return &models.JoinMVResponse{}, nil
+		// Refresh our mvInfo state as per the latest clustermap.
+		// This will undo the changes made by the prev incomplete JoinMV, updating mvInfo as if the
+		// previous JoinMV never happened. If refreshFromClustermap() fails, we cannot safely proceed.
+		// For newMV, we won't have the MV in clustermap yet, so no need to refresh.
+		//
+		if !newMV {
+			err := mvInfo.refreshFromClustermap()
```

Comment:
After this block, we are deleting the mvInfo from mvMap. So, what is the use of refreshing the clustermap here?
```
// Remove the MV replica, we will add a fresh one later down.
rvInfo.deleteFromMVMap(req.MV)
mvInfo = nil
```

Reply:
we want to update rv.reservedSpace and any other "undo" that it may do.. it's best to behave as if we went back to the clustermap state, before applying the new update